### PR TITLE
🐛 Fix JSON Schema for `ValidationError` at field `loc`

### PR DESCRIPTION
**Pull request by dconathan**
See #3790

Basically, Pydantic's schema for the `loc` field of the error detail is `Union[str, int]` not `str` for e.g. cases when the invalid data is at a particular index of an array:

See the type defined in Pydantic's source here:
https://github.com/samuelcolvin/pydantic/blob/5ccbdcb5904f35834300b01432a665c75dc02296/pydantic/error_wrappers.py#L12

The `openapi.json` schema for a `ValidationError` should match this.

This is particularly important for automated testing using https://schemathesis.readthedocs.io/en/stable/ because ValidationErrors that don't match the schema will lead to failing tests.

I've updated the `validation_error_definition` (h/t to @silversurfer34 for creating the issue and supplying the fixed schema 😄 ).
Also I had to update all the tests since the `openapi.json`s are hardcoded in that.

### CODE REVIEW

File: fastapi/openapi/utils.py [hunk 1]
```diff
@@ -37,7 +37,11 @@
     "title": "ValidationError",
     "type": "object",
     "properties": {
-        "loc": {"title": "Location", "type": "array", "items": {"type": "string"}},
+        "loc": {
+            "title": "Location",
+            "type": "array",
+            "items": {"oneOf": [{"type": "string"}, {"type": "integer"}]},
```
Comment (Stranger6667) [inline fastapi/openapi/utils.py:43]: I'd suggest using `anyOf` instead (as Open API 3.0 doesn't support `type` as arrays) - it is slightly faster on the validation side. It doesn't require validating the input instance against all sub-schemas + these types are mutually exclusive (e.g. `number` vs `integer` might give different results under `anyOf` and `oneOf`).

**Review by silversurfer34 — APPROVED** [review]
Looks good to me

File: fastapi/openapi/utils.py (hunk 1 above)
Comment (Stranger6667) [reply fastapi/openapi/utils.py]: Btw, the "Dog vs. Cat" example on that page is wrong :) Neither schema enforces `required` properties or restricts extra ones (via `additionalProperties: false`). E.g. `{"bark": true, "breed": "Dingo"}` is a valid `Cat` instance too!

**Comment by tiangolo** [comment]
Awesome, thanks @dconatha for the work and explanation! :nerd_face: :cake: 

This will be available in FastAPI `0.75.2`, released in a couple of hours. :rocket:

**Pull request merged**